File: py/rt/camera.py
import math
from vec3 import vec3
from ray import ray

class camera:
	def __init__(self, lookFrom, lookAt, vUp, fovy, aspectRatio, aperture, distToFocus):
		self.lookFrom = lookFrom
		self.lookAt = lookAt
		self.vUp = vUp
		self.fovy = math.radians(fovy)
		self.aspectRatio = aspectRatio
		self.lenseRadius = aperture/2
		self.distToFocus = distToFocus
		
		self.halfHeight = math.tan(self.fovy/2) * distToFocus
		self.halfWidth  = self.halfHeight * aspectRatio
		
		self.w = vec3.normalize(lookAt-lookFrom)
		self.u = vec3.normalize(vec3.cross(self.w,vUp))
		self.v = vec3.cross(self.u,self.w)

		self.lowerLeftCorner = lookFrom + self.w*distToFocus - self.v*self.halfHeight - self.u*self.halfWidth		
		
	def getRay(self, u, v):		
		temp = vec3.randomPointInDisc() * self.lenseRadius
		apertureOffset = self.u * temp.x() + self.v * temp.y()
		return ray(self.lookFrom + apertureOffset, self.lowerLeftCorner + self.u*u*2*self.halfWidth + self.v*v*2*self.halfHeight - (self.lookFrom + apertureOffset))

File: cpp/OpenGL/22_Raytrace/Camera.cpp
#include <cmath>

#include "Camera.h"

Camera::Camera(const Vec3& lookFrom, const Vec3& lookAt, const Vec3& vUp, float fovyDeg, 
			   float aspectRatio, float aperture, float distToFocus) :
	lookFrom{lookFrom},
	lookAt{lookAt},
	vUp{vUp},
	fovy{3.14159265f * fovyDeg / 180.0f},
	lenseRadius{aperture/2},
	halfHeight{tanf(fovy/2) * distToFocus},
	halfWidth{halfHeight * aspectRatio},
	w{Vec3::normalize(lookAt-lookFrom)},
	u{Vec3::normalize(Vec3::cross(w,vUp))},
	v{Vec3::cross(u,w)},
	lowerLeftCorner{lookFrom + w*distToFocus - v*halfHeight - u*halfWidth}
{
}

Ray Camera::getRay(const float s, const float t) const {
	const Vec3 temp{Vec3::randomPointInDisc() * lenseRadius};
	const Vec3 apertureOffset{u*temp.x + v*temp.y};
	return Ray{lookFrom + apertureOffset, lowerLeftCorner + u*s*2*halfWidth + v*t*2*halfHeight  - (lookFrom + apertureOffset)};
}
